File: pages/refactorings/replace-inheritance-with-natural-transformation/03-either-option/After.scala
// Order quantity from a raw form field. One canonical type at
// the boundary; each view is an arrow, not a superclass.
object After:
  def parse(raw: String): Either[String, Int] =
    raw.toIntOption match
      case Some(n) if n > 0 => Right(n)
      case Some(_)          => Left("not positive")
      case None             => Left("not a number")

  def toOption[E, A](e: Either[E, A]): Option[A] = // natural in A
    e.fold(_ => None, Some(_))

  def quantity(raw: String): Option[Int] =
    toOption(parse(raw))

File: pages/refactorings/replace-inheritance-with-natural-transformation/03-either-option/After.hs
-- Order quantity from a raw form field. One canonical type at
-- the boundary; each view is an arrow, not a superclass.
module After where

import Text.Read (readMaybe)

parse :: String -> Either String Int
parse raw = case readMaybe raw of
  Just n | n > 0 -> Right n
  Just _         -> Left "not positive"
  Nothing        -> Left "not a number"

toMaybe :: Either e a -> Maybe a -- natural in a
toMaybe = either (const Nothing) Just

quantity :: String -> Maybe Int
quantity raw = toMaybe (parse raw)
